staging/src/k8s.io/client-go/resolver/resolver.go

```diff
@@ -0,0 +1,327 @@
+//   is not an IP address or is resolved via /etc/hosts.
+// - The configured API server URL has a different port
+//   than the one used in the Endpoints.
+func NewResolver(ctx context.Context, c *rest.Config, options ...ResolverOption) (*net.Resolver, error) {
+	config := *c
+	if err := setConfigDefaults(&config); err != nil {
+		return nil, err
+	}
+
+	host, port, err := getHostPort(config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	if netutils.ParseIPSloppy(host) != nil {
+		return net.DefaultResolver, fmt.Errorf("APIServerResolver only works for domain names")
+	}
+
+	// defaulting
+	r := &resolver{
+		host:    host,
+		port:    port,
+		timeout: 100 * time.Second,
+	}
+
+	// options
+	for _, o := range options {
+		o(r)
+	}
+
+	f := &dns.MemResolver{
+		LookupIP: func(ctx context.Context, network, host string) ([]net.IP, error) {
+			return r.lookupIP(ctx, network, host)
+		},
+	}
+
+	resolver := dns.NewMemoryResolver(f)
+	config.Resolver = resolver
+
+	client, err := rest.RESTClientFor(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	r.client = client
+	// Initialize cache and close idle connections so next connections goes
+	// directly to one of the API server IPs, this is useful for cases
+	// that use a load balancer for bootstrapping
+	r.refreshCache(ctx)
+	utilnet.CloseIdleConnectionsFor(r.client.Client.Transport)
+	return resolver, nil
+}
+
+func setConfigDefaults(config *rest.Config) error {
+	gv := v1.SchemeGroupVersion
+	config.GroupVersion = &gv
+	config.APIPath = "/api"
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+
+	if config.UserAgent == "" {
+		config.UserAgent = rest.DefaultKubernetesUserAgent()
+	}
+
+	return nil
+}
+
+func (r *resolver) lookupIP(ctx context.Context, network, host string) ([]net.IP, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// Use the default resolver if is not trying to resolve the configured API server hostname
+	if !strings.HasPrefix(host, r.host) {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, different than API server hostname %s", host, r.host)
+		return net.DefaultResolver.LookupIP(ctx, network, host)
+	}
+
+	// Use the default resolver if the cache is empty and try to renew the cache.
+	if len(r.cache) == 0 {
+		klog.V(7).Infof("Resolver Trace: use default resolver for host %s, cache is empty", host)
+		// recursion is handled by std library that coalesce duplicate in-flight DNS lookups
+		// https://github.com/golang/go/commit/61d3b2db6292581fc07a3767ec23ec94ad6100d1
+
+		// Try to renew the
+		addrs, err := net.DefaultResolver.LookupIP(ctx, network, host)
+		if err != nil {
+			return addrs, err
+		}
+		// don't try to renew the cache until we know it's possible
+		if len(addrs) > 0 {
+			go r.refreshCache(ctx)
+		}
+		return addrs, nil
+	}
+
+	// First query after the cache was created
+	// set the timestamp and try to renew the cache
+	if r.timestamp.IsZero() {
+		r.timestamp = time.Now()
+	}
+	go r.refreshCache(ctx)
```

Comment:
we don't wait?

Reply:
we can't block the resolver, since refreshCache createa a new http request that ask teh resolver, this recursion is the one I need to handle to only allow one depth